File: python/phone-number/phone.py
import re

class Phone:
    def __init__(self, number):
        self._number = number

    @property
    def number(self):
        temp = re.sub(r"[^0-9]", "", self._number)
        if len(temp) != 10:
            if not (len(temp) == 11 and temp[0] == "1"):
                return "0000000000"
            return temp[1:]
        return temp

    def area_code(self):
        return self.number[:3]

    def pretty(self):
        return re.sub(r"([0-9]{3})([0-9]{3})([0-9]{4})", r"(\1) \2-\3", self.number)

File: haskell/phone-number/Phone.hs
module Phone(areaCode, number, prettyPrint) where
import Data.Char(isDigit)

areaCode :: String -> String
areaCode x = take 3 $ number x

number :: String -> String
number x
    | length x' == 10 = x'
    | length x' == 11  && head x' == '1' = tail x'
    | otherwise = replicate 10 '0'
        where x' = filter isDigit x

prettyPrint :: String -> String
prettyPrint x = "(" ++ p1 ++ ") " ++ p2 ++ "-" ++ p3
    where
        (p1,rest) = splitAt 3 (number x)
        (p2,p3) = splitAt 3 rest